plugin/awesoMD/plugin.js

```diff
@@ -454,35 +458,85 @@ const Plugin = () => {
 		return [content, options];
 	}
 
+	/**
+	 * Separates the inline metadata and content for slide having inline metadata in yaml block as
+	 *
+	 * ```yaml
+	 * metadata_key1: metadata_value1
+	 * metadata_key2: metadata_value2
+	 * ```
+	 */
+	function extractYAMLMetadata(markdown, options) {
+		const markdownParts = yamlRegex.exec(markdown)
+		yamlRegex.lastIndex = 0
+		if (markdownParts && markdownParts[2]) {
+			const metadata = markdownParts[2]
+			markdown = markdownParts[3] || ''
+
+			try {
+				const metadataYAML = yaml.load(metadata)
+				if (metadataYAML === undefined) {
+					throw new Error("The inline metadata is not valid.")
+				}
+				options.metadata = { ...options.metadata, ...metadataYAML }
+				options.attributes = 'class=' + (options.metadata.slide || '')
+			} catch (error) {
+				console.error(error)
+				markdown = error.message
+			}
+		}
+		return [markdown, options]
+	}
+
+	/**
+	 * Separates the inline metadata and content for slides having metadata as
+	 *
+	 * ::metadata_key1:metadata_value1 ::metadata_key2:metadata_value2
+	 */
+	function extractInlineMetadata(markdown, options) {
+		const inlineMetadata = {}
+		const matches = markdown.match(headingWithMetadataRegex)
+
+		if (matches && matches[0]) {
+			const metadataMatches = matches[0].match(metadataRegex)
+			if (metadataMatches) {
+				metadataMatches.forEach(metadataMatch => {
+					const [key, value] = metadataMatch.replace('::', '').split(':')
+					inlineMetadata[key.trim()] = value.trim()
+					const metadataPattern = new RegExp(`::\\b${key.trim()}\\b:\\s*${value.trim()}`)
+					markdown = markdown.replace(metadataPattern, '')
+				})
+			}
+		}
+
+		options.metadata = { ...options.metadata, ...inlineMetadata }
+		options.attributes = 'class=' + (options.metadata.slide || '')
+		return [markdown, options]
+	}
+
 	/**
 	 * Separates the inline metadata and content for each slide
 	 *
 	 * Returns updated options with the inline metadata and
 	 * updated markdown without the inline metadata for each slide
 	 */
 	function separateInlineMetadataAndMarkdown(markdown, options) {
-		const yamlRegex =  /```(yaml|yml)\n([\s\S]*?)```(\n[\s\S]*)?/g;
-		if (yamlRegex.test(markdown)){
-			yamlRegex.lastIndex = 0;
-
-			const markdownParts = yamlRegex.exec(markdown)
-			const metadata = markdownParts[2] || {}
-			markdown = markdownParts[3] || ''
-			if (metadata){
-				try {
-					const metadataYAML = yaml.load(metadata);
-					if (metadataYAML === undefined) {
-						throw new Error("The inline metadata is not valid.")
-					}
-					options.metadata = {...options.metadata, ...metadataYAML}
-					options.attributes = 'class=' + options.metadata.slide;
-				} catch (error) {
-					markdown = error.message
-					console.error(error)
+		const yamlMetadata = yamlRegex.test(markdown)
+		const newMetadata = headingWithMetadataRegex.test(markdown)
+		yamlRegex.lastIndex = 0
+
+		switch (true) {
```

Comment:
what is that supposed to do?

Reply:
It will check for the type of metadata provided, either in yaml format or the new inline metadata style metadata. If either one style of metadata is used then the respective variables will be set to true.
`yamlRegex.lastIndex = 0`, this will reset the last index property to search the string from the start since the regex uses the global flag.
And finally using switch case to extract the metadata from the type of style of metadata used in the markdown file.

Reply:
So now this logic will apply:

`newMetadata == true && yamlMetadata == false ` =>  `extractInlineMetadata`
`newMetadata == false && yamlMetadata == true ` =>  `extractYAMLMetadata`
`newMetadata == false && yamlMetadata == false && options.metadata ` =>  `options.attributes = 'class=' + (options.metadata.slide || '')`
`newMetadata == true && yamlMetadata == true ` =>  `extractInlineMetadata`
`newMetadata == false && yamlMetadata == false && !options.metadata ` =>  nothing

is that what we want? 
If both metadata types are available,  the inlineMetadata will be used.
I thought we wanted to use the new switch (see #9) to know which metadata to expect

Reply:
Currently, the plugin supports the following markdown with different metadata:
Type 1: All slides containing yaml metadata
```md
    ```yaml
    slide: cover
    toc: false
    ```
    # Title with line wrapping

    ---

    ```yaml
    slide: section
    ```
    # Overview traditional project managements
```

Type 2: All slides containing new metadata style
```md
    # Title with line wrapping ::slide:cover ::toc:false
    ---
    # Overview traditional project management ::slide:section
```

Type 3: some slide contains yaml metadata and some slide containing new metadata
```md
    ```yaml
    slide: cover
    toc: false
    ```
    # Title with line wrapping
    ---
    # Overview tradition project management ::slide:section
```

These 3 types of markdowns are currently supported by our plugin.

Do you mean our plugin should also support the following type of markdown?

Type 4: one slide that contains both yaml and new metadata type
```md
    ```yaml
    slide: cover
    ```
    # Title with line wrapping ::toc:false
    ---
    # Overview tradition project management ::slide:section
```

With which metadata to use, this cannot be filtered in this PR since we don't have that attribute added in this PR and the filtering part will be added in the other PR https://github.com/JankariTech/revealjs-awesoMD/pull/9. This PR is only for extracting metadata from the slide.

Reply:
> There is an other case:
> 
> ```
> # Title with line wrapping ::slide:cover ::toc:false
> # Overview traditional project management ::slide:section
> ```
> 
> what should happen in that case?

Yes, that's the case that will be covered in the next PR. We don't have an attribute `separate-by-heading` in this PR. For this case the plugin should add the slide separator and this is done in the https://github.com/JankariTech/revealjs-awesoMD/pull/9. This is the reason this case has not been covered in this PR.

Currently we are going with the following flow:
1. Extract metadata for the markdown file which have slide separator in it. (Covered in this PR)
2. Then work for the case where there is no slide separator. (Will be covered in the following PR)

Reply:
> If we want to allow to mix different types of metadata (type 3) that is fine, it's just different to what I've expected and might be confusing.
> 
> I don't think we need to support type 4, but we need to be clear what should happen if the user does that. With this PR, the yaml metadata would be ignored. Will that confuse users?

Yes, mixing different types of metadata doesn't seem to be a good idea and will be confusing to the users. I think we can check whether the same type of metadata is provided in every slide. If not then we can inform user to maintain the consistency on every slide.